### DISCUSSION TITLE
How to embed Twitter widget?

### DISCUSSION THREAD

**Discussion by mikan3rd** [vercel/next.js#15262] [Help] [answered]
I created widget from this website and tried to embed Twitter widget like this
https://publish.twitter.com/

##### _document.tsx
```_document.tsx
import React from "react";
import Document, { Head, Html, Main, NextScript } from "next/document";

class MyDocument extends Document {
  render() {
    return (
      <Html>
        <Head>
          <script async src="https://platform.twitter.com/widgets.js" />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}

export default MyDocument;
```

##### index.tsx
```index.tsx
import React from "react";

const IndexPage: React.FC = () => {
  return (
    <a
      className="twitter-timeline"
      data-height="400"
      href="https://twitter.com/commitly_jp"
      target="_blank"
      rel="noreferrer"
    >
      Tweets by commitly_jp
    </a>
  );
};

export default IndexPage;
```

I can embed Tweet timeline, but the following warning appears in the console
```
react-dom.development.js:88
Warning: Extra attributes from the server: data-twitter-extracted-i15949943201013305
```

Please tell me how to embed it correctly.

Note: This problem only seems to occur if I start with `next dev`.

**Comment by jacargentina** [comment]
Also i'm having a similar error when using a <video> tag like this:

```jsx
      <video
        autoPlay
        loop
        muted
        playsInline
        style={{
          height: '100vh',
          width: '100vw',
          objectFit: 'cover',
          zIndex: 0,
        }}>
        <source src={url} type="video/mp4" />
      </video>
```

The error says in my case `Warning: Extra attributes from the server:  data-video`

How do i solve it?

**Comment by mikan3rd** [comment, marked as answer]
I gave up embedding the script in `_document.tsx` and dynamically plugged it in as follows

```tsx
  React.useEffect(() => {
    const s = document.createElement("script");
    s.setAttribute("src", "https://platform.twitter.com/widgets.js");
    s.setAttribute("async", "true");
    document.head.appendChild(s);
  }, []);
```

**Comment by kaisugi** [comment]
If you don't mind using external modules, [react-twitter-embed](https://github.com/saurabhnemade/react-twitter-embed) is a good way. I use this module in my personal Next-based blog.

Simple Tweet Embedding Example:

```tsx
import { TwitterTweetEmbed } from "react-twitter-embed";

<TwitterTweetEmbed
  tweetId={'340640143058825216'}
/>
```

FYI: https://saurabhnemade.github.io/react-twitter-embed/

**Reply by verheyenkoen** [reply]
Wouldn't that just do everything on the client? With Next.js you could leverage the loading of the oembed code on the server which I guess you're not getting with such library.

**Reply by k4200** [reply]
I tried to use it, but it doesn't work with React 19.

**Comment by iietmoon** [comment]
> I created widget from this website and tried to embed Twitter widget like this https://publish.twitter.com/
> 
> ##### _document.tsx
> ```tsx
> import React from "react";
> import Document, { Head, Html, Main, NextScript } from "next/document";
> 
> class MyDocument extends Document {
>   render() {
>     return (
>       <Html>
>         <Head>
>           <script async src="https://platform.twitter.com/widgets.js" />
>         </Head>
>         <body>
>           <Main />
>           <NextScript />
>         </body>
>       </Html>
>     );
>   }
> }
> 
> export default MyDocument;
> ```
> 
> ##### index.tsx
> ```tsx
> import React from "react";
> 
> const IndexPage: React.FC = () => {
>   return (
>     <a
>       className="twitter-timeline"
>       data-height="400"
>       href="https://twitter.com/commitly_jp"
>       target="_blank"
>       rel="noreferrer"
>     >
>       Tweets by commitly_jp
>     </a>
>   );
> };
> 
> export default IndexPage;
> ```
> 
> I can embed Tweet timeline, but the following warning appears in the console
> 
> ```
> react-dom.development.js:88
> Warning: Extra attributes from the server: data-twitter-extracted-i15949943201013305
> ```
> 
> Please tell me how to embed it correctly.
> 
> Note: This problem only seems to occur if I start with `next dev`.

it worked for me, thanks you

**Comment by aceisScope** [comment]
A side note, embedded twitter timeline just won't work on safari 🤔 maybe due to its privacy policy?

**Reply by ghost** [reply]
privacy policy !?

**Comment by ca-kraa** [comment]
# For Alternative Use This [React Tweet](https://react-tweet.vercel.app/)

React Tweet is a React component that makes it easy to embed tweets from Twitter into your web application.

## Installation

You can install React Tweet using npm:

```bash
npm install react-tweet
```

## Example Code

```javascript
import  { Tweet } from "react-tweet";

function Twitter() {
  return (
    <div className="light">
      <Tweet id="1723990780145184971" />
    </div>
  );
}

export default Twitter;
````
